sdk/lib/opentelemetry/sdk/trace/span.rb

```diff
@@ -118,6 +118,37 @@ def add_attributes(attributes)
           self
         end
 
+        # Add a link to a {Span}.
+        #
+        # Adding links at span creation using the `links` option is preferred
+        # to calling add_link later, because head sampling decisions can only
+        # consider information present during span creation.
+        #
+        # Example:
+        #
+        #   span.add_link(OpenTelemetry::Trace::Link.new(span_to_link_from.context))
+        #
+        # Note that the OpenTelemetry project
+        # {https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-semantic-conventions.md
+        # documents} certain "standard attributes" that have prescribed semantic
+        # meanings.
+        #
+        # @param [OpenTelemetry::Trace::Link] the link object to add on the {Span}.
+        #
+        # @return [self] returns itself
+        def add_link(link)
+          @mutex.synchronize do
+            if @ended
+              OpenTelemetry.logger.warn('Calling add_link on an ended Span.')
+            else
+              @links ||= []
+              @links = trim_links(@links + [link], @span_limits.link_count_limit, @span_limits.link_attribute_count_limit)
```

Comment:
`@links + [link]` will allocate 2 additional arrays (one to wrap `link` and one for the `+` result). It should be more efficient to `@links << link`, which will append to the `@links` array in place. The potentially inconsistent intermediate state should be invisible to observers due to the mutex.

Reply:
The links array is frozen in `trim_links`. So in order to do that, I'd have to stop freezing it.

Reply:
yes - stop freezing it because it is now mutable (it wasn't before). It should be frozen in `finish` instead.